src/systems/elevator/state_machine/ElevatorStateMachineImpl.hh

```diff
@@ -0,0 +1,141 @@
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE
+{
+namespace systems
+{
+class ElevatorStateMachinePrivate
+{
+  /// \brief Constructor
+  /// \param[in] _system Data of the enclosing system
+  public: ElevatorStateMachinePrivate(
+      const std::shared_ptr<ElevatorCommonPrivate> &_system);
+
+  /// \brief Adds a new target to the queue
+  /// \details If there is another target in the queue, and the new target is
+  /// an intermediate stop, the new target will take precedence
+  /// \param[in] _target Target to add to the queue
+  public: void EnqueueNewTarget(double _target);
+
+  /// \brief Publishes a command message
+  /// \param[in] _pub Joint command publisher
+  /// \param[in] _cmd Joint command
+  public: void SendCmd(transport::Node::Publisher &_pub, double _cmd);
+
+  /// \brief Data of the enclosing system
+  public: std::shared_ptr<ElevatorCommonPrivate> system;
+
+  /// \brief Floor target queue
+  public: std::deque<int32_t> targets;
+};
+
+//////////////////////////////////////////////////
+ElevatorStateMachinePrivate::ElevatorStateMachinePrivate(
+    const std::shared_ptr<ElevatorCommonPrivate> &_system)
+    : system(_system)
+{
+}
+
+//////////////////////////////////////////////////
+void ElevatorStateMachinePrivate::EnqueueNewTarget(double _target)
+{
+  // Ignore duplicate targets
+  auto it = std::find(this->targets.cbegin(), this->targets.cend(), _target);
+  if (it != this->targets.cend()) return;
+
+  // Prioritize target in the queue
+  bool enqueued = false;
+  int32_t prevTarget = this->system->state;
+  for (it = this->targets.cbegin(); it != this->targets.cend(); ++it)
+  {
+    int32_t lower = prevTarget;
+    int32_t upper = *it;
+    if (upper < lower) std::swap(lower, upper);
+    if (_target >= lower && _target < upper)
+    {
+      this->targets.insert(it, _target);
+      enqueued = true;
+      break;
+    }
+    prevTarget = *it;
+  }
+  if (!enqueued) this->targets.push_back(_target);
+
+  std::ostringstream ss;
+  ss << "The elevator enqueued target " << _target << " [ ";
+  for (const auto &target : this->targets) ss << target << " ";
+  ignmsg << ss.str() << "]" << std::endl;
+}
+
+//////////////////////////////////////////////////
+void ElevatorStateMachinePrivate::SendCmd(transport::Node::Publisher &_pub,
+                                          double _cmd)
+{
+  msgs::Double msg;
+  msg.set_data(_cmd);
+  _pub.Publish(msg);
+}
+
+//////////////////////////////////////////////////
+ElevatorStateMachineDef::ElevatorStateMachineDef(
+    const std::shared_ptr<ElevatorCommonPrivate> &_system)
+    : dataPtr(std::make_unique<ElevatorStateMachinePrivate>(_system))
+{
+}
+
+//////////////////////////////////////////////////
+ElevatorStateMachineDef::~ElevatorStateMachineDef() = default;
+
+//////////////////////////////////////////////////
+const std::unique_ptr<ElevatorStateMachinePrivate>
+    &ElevatorStateMachineDef::Data() const
+{
+  return this->dataPtr;
```

Comment:
I think this defeats the purpose of using the PIMPL pattern. If other classes need to use this data, then it shouldn't be private, or there should be setters and getters for it.

Reply:
I'll make it public. In this case, I didn't have PIMPL in mind. I used dataPtr so I could pass data to the actions/guards/states.

Beyond that, I used PIMPL in all classes because the documentation says so. But does it serve any purpose here? All classes are internal to the elevator system, and they are compiled together in one library.